libcob/profiling.c

```diff
@@ -0,0 +1,445 @@
+   along with GnuCOBOL.  If not, see <https://www.gnu.org/licenses/>.
+*/
+
+#include <limits.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <time.h>
+
+/* include internal and external libcob definitions, forcing exports */
+#define COB_LIB_EXPIMP
+#include "coblocal.h"
+
+#include "tarstamp.h"
+#include "config.h"
+#include "common.h"
+#include "cobprof.h"
+
+#include <sys/types.h>
+#ifdef HAVE_UNISTD_H
+#include <unistd.h>
+#endif
+
+#ifdef	_WIN32
+#define WIN32_LEAN_AND_MEAN
+#include <windows.h>
+#undef MOUSE_MOVED
+#include <process.h>
+#include <io.h>
+#include <fcntl.h>	/* for _O_BINARY only */
+#endif
+
+/* Local types and variables */
+
+struct cobprof_info_list {
+	struct cobprof_info *info ;
+	struct cobprof_info_list *next;
+};
+
+static struct cobprof_info_list *prof_info_list ;
+
+#define PROF_DEFAULT_DEPTH 255
+#define PROF_MAX_DEPTH 10000000
+static int max_prof_depth = PROF_DEFAULT_DEPTH;
+static cob_ns_time          *start_times;
+static int                  *called_procedures;
+static struct cobprof_info* *called_runtimes;
+static int current_idx = -1;
+static int is_active = 1; /* we may want to disable profiling globally */
+static int is_test = 0;
+static const char* prof_program_id;
+
+static cob_global               *cobglobptr = NULL;
+static cob_settings             *cobsetptr = NULL;
+
+
+
+
+static cob_ns_time
+get_ns_time (void)
+{
+	static cob_ns_time ns_time = 0;
+
+	unsigned long long nanoseconds;
+#if defined (_WIN32)
+	LARGE_INTEGER performance_counter;
+	LARGE_INTEGER performance_frequency;
+#elif defined (HAVE_CLOCK_GETTIME)
+	struct timespec ts;
+#else
+	clock_t c;
+#endif
+
+#if defined (_WIN32)
+	QueryPerformanceCounter(&performance_counter);
+	QueryPerformanceFrequency(&performance_frequency);
+	performance_counter.QuadPart *= 1000000000;
+	performance_counter.QuadPart /= performance_frequency.QuadPart;
+	nanoseconds = performance_counter.QuadPart;
+#elif defined (HAVE_CLOCK_GETTIME)
+	clock_gettime(CLOCK_MONOTONIC, &ts);
+	nanoseconds = ts.tv_sec * 1000000000 + ts.tv_nsec;
+#else
+	c = clock();
+	nanoseconds = c * 1000000000;
+	nanoseconds /= CLOCKS_PER_SEC;
+#endif
+
+	if (nanoseconds>ns_time) ns_time = nanoseconds;
+	return ns_time;
+}
+
+static int
+is_in_same_section (struct cobprof_info *info, int proc_idx1, int proc_idx2)
+{
+	int i = 0;
+	const char* name1 = info->procedures_names[proc_idx1];
+	const char* name2 = info->procedures_names[proc_idx2];
+
+	while (name1[i]){
+		if (name1[i] != name2[i]) return 0;
+		if (name1[i] == '|') return 1;
+		i++;
+	}
+	return 0; /* should never happen */
+}
+
+static int
+is_section (struct cobprof_info *info, int proc_idx)
+{
+    return !!strstr (info->procedures_names[proc_idx], "||");
+}
+
+void cob_init_prof (cob_global *lptr, cob_settings *sptr)
+{
+	cobglobptr = lptr;
+        cobsetptr  = sptr;
+
+	is_test = cobsetptr->cob_testsuite_mode;
+	is_active = !cobsetptr->cob_prof_disable;
```

Comment:
I don't understand the case of static variables. `cob_init_prof` is only called once, after reading the settings, so, why would we need to call it again ? Does it mean that the config could be modified during the execution, and so the static variables should be updated every time ?

Reply:
Note that a solution to the config being modified during execution would be to delay the modification of static variables until the first call to `cob_prof_init`, i.e. the execution of the first module to profile. This way, the config can be modified as many times as necessary before, but any later modification will not impact the profiling (as we probably want).

Reply:
Yes, they can be adjusted during runtime, in this case to switch profiling on/off.

In production you often have to tackle a set of programs and can't do anything else about the rest, so enabling/disabling in the COBOL code can be quite useful, especially if you have "init/teardown code" that you're not interested in profiling (because you can't do anything about that).

As noted: check the screenio handling for settings (common.c pushes interesting seeing changes down to the libcob module).

For the testsuite mode: that's really intern and static, a simple getenv where you need it is enough.